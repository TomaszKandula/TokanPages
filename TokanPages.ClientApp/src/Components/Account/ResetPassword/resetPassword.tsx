import * as React from "react";
import { useDispatch, useSelector } from "react-redux";
import { ApplicationState } from "../../../Store/Configuration";
import { OperationStatus } from "../../../Shared/enums";
import { ReactChangeEvent, ReactKeyboardEvent } from "../../../Shared/types";
import { ResetPasswordView } from "./View/resetPasswordView";
import Validate from "validate.js";

import { ApplicationDialogAction, UserPasswordResetAction } from "../../../Store/Actions";

import { ResetFormInput, ValidateResetForm } from "../../../Shared/Services/FormValidation";

import { GetTextWarning, SuccessMessage, WarningMessage } from "../../../Shared/Services/Utilities";

import { RECEIVED_ERROR_MESSAGE } from "../../../Shared/constants";

const formDefaultValues: ResetFormInput = {
    email: "",
};

export const ResetPassword = (): JSX.Element => {
    const dispatch = useDispatch();

    const template = useSelector((state: ApplicationState) => state.contentTemplates?.content); 
    const content = useSelector((state: ApplicationState) => state.contentResetPassword);
    const reset = useSelector((state: ApplicationState) => state.userPasswordReset);
    const error = useSelector((state: ApplicationState) => state.applicationError);

    const hasNotStarted = reset?.status === OperationStatus.notStarted;
    const hasFinished = reset?.status === OperationStatus.hasFinished;
    const hasError = error?.errorMessage === RECEIVED_ERROR_MESSAGE;

    const showSuccess = (text: string) => dispatch(ApplicationDialogAction.raise(SuccessMessage(template.forms.textPasswordReset, text)));
    const showWarning = (text: string) => dispatch(ApplicationDialogAction.raise(WarningMessage(template.forms.textPasswordReset, text)));

    const [form, setForm] = React.useState(formDefaultValues);
    const [hasProgress, setHasProgress] = React.useState(false);

    const clearForm = React.useCallback(() => {
        if (!hasProgress) return;
        setHasProgress(false);
        dispatch(UserPasswordResetAction.clear());
    }, [hasProgress]);

    React.useEffect(() => {
        if (hasError) {
            clearForm();
            return;
        }

        if (hasNotStarted && hasProgress) {
            dispatch(
                UserPasswordResetAction.reset({
                    emailAddress: form.email,
                })
            );

            return;
        }

        if (hasFinished) {
            clearForm();
            setForm(formDefaultValues);
            showSuccess(template.templates.password.resetSuccess);
        }
    }, [hasProgress, hasError, hasNotStarted, hasFinished, template]);

    const keyHandler = React.useCallback((event: ReactKeyboardEvent) => {
        if (event.code === "Enter") {
            buttonHandler();
        }
    }, []);

    const formHandler = React.useCallback(
        (event: ReactChangeEvent) => {
            setForm({ ...form, [event.currentTarget.name]: event.currentTarget.value });
        },
        [form]
    );

    const buttonHandler = React.useCallback(() => {
        let results = ValidateResetForm({ email: form.email });

        if (!Validate.isDefined(results)) {
            setHasProgress(true);
            return;
        }

        showWarning(GetTextWarning({ object: results, template: template.templates.password.resetWarning }));
    }, [form, template]);

    return (
        <ResetPasswordView
            isLoading={content?.isLoading}
            progress={hasProgress}
            caption={content?.content?.caption}
            button={content?.content?.button}
            email={form.email}
            keyHandler={keyHandler}
            formHandler={formHandler}
            buttonHandler={buttonHandler}
            labelEmail={content?.content?.labelEmail}
        />
    );
};
