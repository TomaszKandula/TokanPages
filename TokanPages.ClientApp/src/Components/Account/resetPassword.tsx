import * as React from "react";
import { useDispatch, useSelector } from "react-redux";
import { IApplicationState } from "../../Redux/applicationState";
import { IGetResetPasswordContent } from "../../Redux/States/Content/getResetPasswordContentState";
import { ActionCreators as DialogAction } from "../../Redux/Actions/raiseDialogAction";
import WarningMessage from "../../Shared/Components/ApplicationDialogBox/Helpers/warningMessage";
import { ValidateEmail } from "../../Shared/validate";
import { ProduceWarningText } from "../../Shared/textWrappers";
import { RESET_FORM, RESET_PASSWORD_WARNING } from "../../Shared/constants";
import ResetPasswordView from "./resetPasswordView";
import Validate from "validate.js";

const ResetPassword = (props: IGetResetPasswordContent): JSX.Element =>
{
    const dispatch = useDispatch();
    const raiseErrorState = useSelector((state: IApplicationState) => state.raiseError);
    
    const showWarning = React.useCallback((text: string) => dispatch(DialogAction.raiseDialog(WarningMessage(RESET_FORM, text))), [ dispatch ]);

    const [form, setForm] = React.useState({email: ""});
    const [progress, setProgress] = React.useState(false);

    // TODO: add action call...

    const formHandler = (event: React.ChangeEvent<HTMLInputElement>) => 
    { 
        setForm({ ...form, [event.currentTarget.name]: event.currentTarget.value }); 
    };

    const buttonHandler = () =>
    {
        let results = ValidateEmail(form.email);

        if (!Validate.isDefined(results))
        {
            setProgress(true);
            return;
        }

        showWarning(ProduceWarningText(results, RESET_PASSWORD_WARNING));
    };

    return (
        <ResetPasswordView bind=
        {{
            isLoading: props.isLoading,
            progress: progress,
            caption: props.content.caption,
            button: props.content.button,
            formHandler: formHandler,
            buttonHandler: buttonHandler
        }}/>
    );
}

export default ResetPassword;
