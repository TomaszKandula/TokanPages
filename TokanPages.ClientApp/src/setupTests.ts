// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import enableHooks from "jest-react-hooks-shallow";

enableHooks(jest);

jest.mock("react-redux", () => (
{
    useSelector: jest.fn(),
    useDispatch: jest.fn()
}));

jest.mock("connected-react-router", () => (
{
    connectRouter: jest.fn(),
    routerMiddleware: jest.fn()
}));

Enzyme.configure({ adapter: new Adapter(), disableLifecycleMethods: false });

// environmental variables used during all tests
process.env.REACT_APP_API_VER="1";
process.env.REACT_APP_BACKEND="http://localhost:5000";
