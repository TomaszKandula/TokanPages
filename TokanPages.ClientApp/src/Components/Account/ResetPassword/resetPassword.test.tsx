import "../../../setupTests";
import React from "react";
import * as Redux from "react-redux";
import * as Router from "react-router";
import { BrowserRouter } from "react-router-dom";
import { render } from "@testing-library/react";
import { ResetPassword } from "./resetPassword";

jest.mock("react-router", () => ({
    ...(jest.requireActual("react-router") as typeof Router),
}));

jest.mock("react-redux", () => ({
    ...jest.requireActual("react-redux"),
    useSelector: jest.fn(),
}));

describe("test account group component: resetPassword", () => {
    const testContent = {
        language: "eng",
        caption: "Reset Password",
        button: "Reset",
        labelEmail: "Email address",
    };

    const useDispatchMock = jest.spyOn(Redux, "useDispatch");
    beforeEach(() => {
        jest.spyOn(Redux, "useSelector").mockReturnValueOnce({
            isLoading: false,
            content: testContent,
        });

        useDispatchMock.mockReturnValue(jest.fn());
    });

    it("should renders correctly '<ResetPassword />' when content is loaded.", () => {
        const html = render(
            <BrowserRouter>
                <ResetPassword />
            </BrowserRouter>
        );
        expect(useDispatchMock).toBeCalledTimes(1);
        expect(html).toMatchSnapshot();
    });
});
