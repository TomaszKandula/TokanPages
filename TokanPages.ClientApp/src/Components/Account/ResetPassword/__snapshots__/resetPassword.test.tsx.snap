// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test account group component: resetPassword should renders correctly '<ResetPassword />' when content is loaded. 1`] = `
<section
  class="makeStyles-section-1"
>
  <div
    class="MuiContainer-root makeStyles-container-2 MuiContainer-maxWidthLg"
  >
    <div
      class="MuiBox-root MuiBox-root-8"
    >
      <div
        class="MuiPaper-root MuiCard-root makeStyles-card-5 MuiPaper-elevation0 MuiPaper-rounded"
      >
        <div
          class="MuiCardContent-root makeStyles-card_content-6"
        >
          <div
            class="MuiBox-root MuiBox-root-9"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-account-3"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
              />
            </svg>
            <p
              class="MuiTypography-root makeStyles-caption-4 MuiTypography-body1"
            />
          </div>
          <div
            class="MuiBox-root MuiBox-root-10"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
              >
                <div
                  class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
                >
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
                  >
                    <input
                      aria-invalid="false"
                      autocomplete="email"
                      class="MuiInputBase-input MuiOutlinedInput-input"
                      id="email"
                      name="email"
                      required=""
                      type="text"
                      value=""
                    />
                    <fieldset
                      aria-hidden="true"
                      class="PrivateNotchedOutline-root-11 MuiOutlinedInput-notchedOutline"
                      style="padding-left:8px"
                    >
                      <legend
                        class="PrivateNotchedOutline-legend-12"
                        style="width:0.01px"
                      >
                        <span>
                          â€‹
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-15"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-22 MuiButton-fullWidth"
                tabindex="0"
                type="submit"
              >
                <span
                  class="MuiButton-label"
                />
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;
